// microsoft-text-translator/lib/ms-text-translator.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = require("axios");
const qs = require("qs");
const uuidv4 = require("uuid/v4");
const BASE_URL = 'https://api.cognitive.microsofttranslator.com';
class MsTextTranslator {
    constructor(config) {
        this.config = config;
        this.api = axios_1.default.create({
            baseURL: BASE_URL,
            params: {
                'api-version': '3.0',
            },
            headers: {
                'Ocp-Apim-Subscription-Key': this.config.credentials.subscriptionKey,
                'Content-Type': 'application/json',
            },
        });
    }
    getLanguages(scopes = ['translation', 'transliteration', 'dictionary']) {
        return __awaiter(this, void 0, void 0, function* () {
            let params = {};
            if (scopes.length) {
                params = Object.assign(Object.assign({}, params), { scope: scopes.join(',') });
            }
            const response = yield this.api.get('languages', {
                params,
                headers: this.makeHeaders(),
            });
            return response.data;
        });
    }
    translate(data, options) {
        return this.makeRequest('translate', data, options);
    }
    transliterate(data, options) {
        return this.makeRequest('transliterate', data, options);
    }
    detectLanguage(data) {
        return this.makeRequest('detect', data);
    }
    breakSentence(data, options = {}) {
        return this.makeRequest('breaksentence', data, options);
    }
    lookupDictionary(data, options) {
        return this.makeRequest('dictionary/lookup', data, options);
    }
    dictionaryExamples(data, options) {
        return this.makeRequest('dictionary/examples', data, options);
    }
    makeRequest(op, data, params = {}, request = this.api.post) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield request(op, data, {
                params,
                headers: this.makeHeaders(),
                paramsSerializer: function (p) {
                    return qs.stringify(p, { arrayFormat: 'repeat' });
                },
            });
            return response.data;
        });
    }
    makeHeaders() {
        return {
            'X-ClientTraceId': uuidv4().toString(),
        };
    }
}
exports.MsTextTranslator = MsTextTranslator;
